jupyterhub/templates/scheduling/_scheduling-helpers.tpl

```diff
@@ -0,0 +1,47 @@
+{{- /*
+  jupyterhub.prepareScope
+    Requires the .podKind field.
+
+    prepareScope sets fields on the scope to be used by the
+    jupyterhub.tolerations and jupyterhub.affinity helm template functions
+    below.
+*/}}
+{{- define "jupyterhub.prepareScope" -}}
+{{- /* Update the copied scope */}}
+{{- $dummy := set . "component" (include "jupyterhub.componentLabel" .) }}
+
+{{- /* Fetch relevant information */}}
+{{- $dummy := set . "tolerationsList" (include "jupyterhub.tolerationsList" .) }}
+{{- $dummy := set . "tolerations" (include "jupyterhub.tolerations" .) }}
+{{- $dummy := set . "hasTolerations" .tolerations }}
+{{- end }}
+
+
+
+{{- /*
+  jupyterhub.tolerations
+    Refactors the setting of the user pods tolerations field.
+*/}}
+{{- define "jupyterhub.tolerations" -}}
```

Comment:
I think this helper is at least one too far. The define/end is more than the code it's deduplicating by being reused.

Reply:
If you would add

```yaml
tolerations:
```
without any entries within it, you get invalid YAML. To know if you should add the `tolerations:` entry or not, you must first figure out if you have content within it.

`tolerationsList` is the content for the `tolerations`. It would have been a better name with `tolerationsContent`.

A lot of the struggles I've faced relates to "should i add this or not, do i have content for the field or not?", the prepareScope helper takes care of a lot of those questions.

I have tried to make do with less helpers and It is really hard... =/ I could try more and more, but I'm currently out of ideas on how to solve the challenges. I'll try address things better in a single point to avoid spreading discussion.